x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CacheFile.java

```diff
@@ -374,6 +356,27 @@ public void onFailure(Exception e) {
         }
     }
 
+    /**
+     * Get the reference to the currently open file channel for this cache file for a read operation
+     *
+     * @param future completable future to complete after the read operation finished
+     * @return file channel reference
+     */
+    private FileChannelReference getFileChannelReference(CompletableFuture<Integer> future) {
+        final FileChannelReference reference;
+        synchronized (listeners) {
+            ensureOpen();
+            reference = channelRef;
+            assert reference.refCount() > 0 : "impossible to run into a fully released channel reference under the listeners mutex";
+            reference.incRef();
+            future.handle((res, t) -> {
```

Comment:
> I wonder if it was more intuitive to "just" ensure we decRef the FileChannelReference appropriately in the two call-sites?

Yea it's quite annoying to code this up (just adds a lot of random code) actually but let's do it for correctness :)

Reply:
Refactored this in d5eb341a775 now to manually release the ref instead of using the future. Let me know what you think. I was able to keep it reasonably dry and clean I hope :)